src/Microsoft.DotNet.Darc/src/Darc/Operations/GatherDropOperation.cs

```diff
@@ -1601,30 +1600,30 @@ private static bool IsAzureDevOpsArtifactsUrl(string location)
                     Directory.CreateDirectory(directory);
                 }
 
-                if (authHeader != null)
-                {
-                    requestMessage = new HttpRequestMessage(HttpMethod.Get, sourceUri)
-                    {
-                        Headers =
-                        {
-                            { HttpRequestHeader.Authorization.ToString(), authHeader.ToString() }
-                        }
-                    };
-                }
-                else
-                {
-                    requestMessage = new HttpRequestMessage(HttpMethod.Get, sourceUri);
-                }
-
                 // Ensure the parent target directory has been created.
                 using (FileStream outStream = new FileStream(temporaryFileName,
                                                       _options.Overwrite ? FileMode.Create : FileMode.CreateNew,
                                                       FileAccess.Write))
                 {
-                    using (var response = await client.SendAsync(requestMessage))
-                    {
-                        response.EnsureSuccessStatusCode();
+                    using (var response = await ExponentialRetry.Default.RetryAsync(
+                        async () =>
+                        {
+                            using (var requestMessage = new HttpRequestMessage(HttpMethod.Get, sourceUri))
+                            {
+                                if (authHeader != null)
+                                {
+                                    requestMessage.Headers.Authorization = authHeader;
+                                }
+
+                                var response = await client.SendAsync(requestMessage);
+                                response.EnsureSuccessStatusCode();
 
+                                return response;
```

Comment:
Something needs to dispose of these in failure cases, since they won't be returned to the user. :-( Unfortunately, "Response" objects leaking can lead to wonky networking behavior, because they hold a lot of external resources that the garbage collector doesn't "see", so they don't apply memory pressure, and therefore don't trigger collections, so can lead to deadlocks.

This could lead to a really unpleasant bug where we only deadlock occasionally, when there is network instability.  So instead of failing, like it did before, this operation just hangs forever.

Retrying network requests is really finicky, it seems like it would be a good idea to have a method that handles all the intricacies (like disposing of things) so that it could be separated from the "business logic" that is here in this method.  Maybe "ExponentialRetry" could have a "RetryHttp" method on it?

Reply:
In this case, should I just try to use HttpRequestManager(https://github.com/dotnet/arcade-services/blob/main/src/Microsoft.DotNet.Darc/src/DarcLib/Helpers/HttpRequestManager.cs) instead?

Reply:
It doesn't do the dispose correctly either. But if we fixed it there, then maybe? :-)

Reply:
Fixed by adding a dispose to the catch in HttpRequestManger and changing GatherDrop to use the HttpRequestManager.